frontend_tests/node_tests/filter.js

```diff
@@ -715,6 +718,62 @@ run_test('predicate_basics', () => {
         display_recipient: [{id: steve.user_id}, {id: me.user_id}],
     }));
     assert(!predicate({type: 'stream'}));
+
+    const img_msg = {
+        content: `<p><a href="/user_uploads/randompath/test.jpeg">test.jpeg</a></p><div class="message_inline_image"><a href="/user_uploads/randompath/test.jpeg" title="test.jpeg"><img src="/user_uploads/randompath/test.jpeg"></a></div>`,
+    };
+
+    const link_msg = {
+        content: `<p><a href="http://chat.zulip.org">chat.zulip.org</a></p>`,
+    };
+
+    const non_img_attachment_msg = {
+        content: `<p><a href="/user_uploads/randompath/attachemnt.ext">attachment.ext</a></p>`,
+    };
+
+    const no_has_filter_matching_msg = {
+        content: "<p>Testing</p>",
+    };
+
+    predicate = get_predicate([['has', 'non_valid_operand']]);
+    assert(!predicate(img_msg));
+    assert(!predicate(non_img_attachment_msg));
+    assert(!predicate(link_msg));
+    assert(!predicate(no_has_filter_matching_msg));
+
+    // HTML content of message is used to determine if image have link, image or attachment.
+    // We are using jquery to parse the html and find existence of relevant tags/elements.
+    // In tests we need to stub the calls to jquery so using zjquery's .set_find_results method.
+    predicate = get_predicate([['has', 'link']]);
+    $(img_msg.content).set_find_results("a", [$("<a>")]);
+    assert(predicate(img_msg));
+    $(non_img_attachment_msg.content).set_find_results("a", [$("<a>")]);
+    assert(predicate(non_img_attachment_msg));
+    $(link_msg.content).set_find_results("a", [$("<a>")]);
+    assert(predicate(link_msg));
+    $(no_has_filter_matching_msg.content).set_find_results("a", false);
+    assert(!predicate(no_has_filter_matching_msg));
+
+    predicate = get_predicate([['has', 'attachment']]);
+    $(img_msg.content).set_find_results("a[href^='/user_uploads']", [$("<a>")]);
+    assert(predicate(img_msg));
+    $(non_img_attachment_msg.content).set_find_results("a[href^='/user_uploads']", [$("<a>")]);
+    assert(predicate(non_img_attachment_msg));
+    $(link_msg.content).set_find_results("a[href^='/user_uploads']", false);
+    assert(!predicate(link_msg));
+    $(no_has_filter_matching_msg.content).set_find_results("a[href^='/user_uploads']", false);
+    assert(!predicate(no_has_filter_matching_msg));
+
+    predicate = get_predicate([['has', 'image']]);
+    $(img_msg.content).set_find_results(".message_inline_image", [$("<img>")]);
+    assert(predicate(img_msg));
+    $(non_img_attachment_msg.content).set_find_results(".message_inline_image", false);
+    assert(!predicate(non_img_attachment_msg));
+    $(link_msg.content).set_find_results(".message_inline_image", false);
+    assert(!predicate(link_msg));
+    $(no_has_filter_matching_msg.content).set_find_results(".message_inline_image", false);
+    assert(!predicate(no_has_filter_matching_msg));
```

Comment:
I think this test might be more readable if saved the predicates like this:
```
check_has_link = get_predicate([['has', 'link']]);
...
```

And then setup 6 or so messages and just assertions like this:
```
assert(has_link(img_msg))
assert(has_attachment(img_msg))
assert(has_image(img_msg))
```